tests/python/gstaichi/lang/fast_caching/test_args_hasher.py

```diff
@@ -0,0 +1,243 @@
+import dataclasses
+
+import numpy as np
+
+import gstaichi as ti
+from gstaichi.lang._fast_caching import FIELD_METADATA_CACHE_VALUE, args_hasher
+
+from tests import test_utils
+
+
+@test_utils.test()
+def test_args_hasher_numeric() -> None:
+    seen = set()
+    for arg in [3, 5.3, np.int32(3), np.int64(5), np.float32(2), np.float64(2)]:
+        for it in [0, 1]:
+            hash = args_hasher.hash_args([arg])
+            assert hash is not None
+            if it == 0:
+                assert hash not in seen
+                seen.add(hash)
+            else:
+                assert hash in seen
+
+
+@test_utils.test()
+def test_args_hasher_unsupported_data_oriented() -> None:
+    @ti.data_oriented
+    class Foo: ...
+
+    foo = Foo()
+    assert args_hasher.hash_args([foo]) is None
+
+
+@test_utils.test()
+def test_args_hasher_ndarray() -> None:
+    seen = set()
+    for dtype in [ti.i32, ti.i64, ti.f32, ti.f64]:
+        for ndim in [0, 1, 2]:
+            arg = ti.ndarray(dtype, [1] * ndim)
+            for it in [0, 1]:
+                hash = args_hasher.hash_args([arg])
+                assert hash is not None
+                if it == 0:
+                    assert hash not in seen
+                    seen.add(hash)
+                else:
+                    assert hash in seen
+
+
+@test_utils.test()
+def test_args_hasher_ndarray_vector() -> None:
+    seen = set()
+    for dtype in [ti.i32, ti.i64, ti.f32, ti.f64]:
+        for vector_size in [2, 3]:
+            for ndim in [0, 1, 2]:
+                arg = ti.Vector.ndarray(vector_size, dtype, [1] * ndim)
+                for it in [0, 1]:
+                    hash = args_hasher.hash_args([arg])
+                    assert hash is not None
+                    if it == 0:
+                        assert hash not in seen
+                        seen.add(hash)
+                    else:
+                        assert hash in seen
+
+
+@test_utils.test()
+def test_args_hasher_ndarray_matrix() -> None:
+    seen = set()
+    for dtype in [ti.i32, ti.i64, ti.f32, ti.f64]:
+        for m in [2, 3]:
+            for n in [2, 3]:
+                for ndim in [0, 1, 2]:
+                    arg = ti.Matrix.ndarray(m, n, dtype, [1] * ndim)
+                    for it in [0, 1]:
+                        hash = args_hasher.hash_args([arg])
+                        assert hash is not None
+                        if it == 0:
+                            assert hash not in seen
+                            seen.add(hash)
+                        else:
+                            assert hash in seen
```

Comment:
What about defining some helper to share the logic ? it is always the same.

You could just use `itertools.product(*a)` to generate you sequence of combinations you want to test, and pass it to some helper function that would do the loop and keep track of `seen`.